luxonis_ml/nn_archive/config_building_blocks/base_models/input.py

```diff
@@ -60,6 +63,9 @@ class Input(CustomBaseModel):
         description="Data type of the input data (e.g., 'float32')."
     )
     input_type: InputType = Field(description="Type of input data (e.g., 'image').")
+    image_layout: ImageLayout = Field(
```

Comment:
Would it make sense to assume `"chw"` by default?

Reply:
We could go just with "layout".
I do think in most cases it will be an image input, so it would make sense doing a "BCHW" layout by default? I think it won't error out here, but it will fail downstream (e.g. at conversion) where we could potentially catch that and warn the user?

Not sure if we'd want to introduce other dependencies there, but we could partially validate this when generating the NN archive by reading ONNX or TFLITE input shapes?

Reply:
> I think it won't error out here, but it will fail downstream (e.g. at conversion) where we could potentially catch that and warn the user?
That's true, I guess we don't have to worry about every possible mistake of the user (and rather warn them later if they did something wrong). I'm adding the default layout in [217a33f](https://github.com/luxonis/luxonis-ml/pull/109/commits/217a33f4cfefbe733ed0ca5ac3fd2bade2c56339)

Reply:
> we could partially validate this when generating the NN archive by reading ONNX or TFLITE input shapes?

How would you go about that? The input shape doesn't provide information on HWC/CHW.

Reply:
> We could go just with "layout".

I named it `image_layout` intentionally to make things less confusing when adding a model with `InputType.RAW` with input layout different from the HWC/CHW. If we rename it to "layout", would make sense to add additional values to the ImageLayout enum that would be more suitable for the `InputType.RAW` cases (e.g. "custom")?

Reply:
> How would you go about that? The input shape doesn't provide information on HWC/CHW.

I think mainly on ranks/dimensions and then we could have some checks which issue warnings if channel is not significantly larger or smaller than HW.

I understand intention behind `image_layout`, and based on this conversation we would have to add more options. Alternatively, we don't treat it as enum and rather a string. Or a dedicated object for which we validate certain configurations (e.g. order of C against H,W; multiple repetitions of H are not allowed; ...).

Reply:
> I think mainly on ranks/dimensions and then we could have some checks which issue warnings if channel is not significantly larger or smaller than HW.

If I understand correctly, we could simply add an additional model_validator to the Input class doing the checks on `image_layout` and `shape` (no need to validate this when generating the NN archive).

Reply:
> I understand intention behind image_layout, and based on this conversation we would have to add more options. Alternatively, we don't treat it as enum and rather a string. Or a dedicated object for which we validate certain configurations (e.g. order of C against H,W; multiple repetitions of H are not allowed; ...).

What more options do you have in mind? Also, I'm not sure I get the the dedicated object part.

Reply:
I've switched to textual layout in [d5476aa](https://github.com/luxonis/luxonis-ml/pull/109/commits/d5476aa083eac8d602a1ad898f0f4ca731ecadad). For validation, I'd still propose to do it with model_validator inside the `Input` instead of at the time of nn archive creation. In my logic, user can now input something random as layout, e.g. "skoQSKO/?!QSqK", which makes the config invalid (OR is there some reason why we should allow this and only fail later?). I'd include the following tests:
- fail if len(layout) =! len(shape)
- fail if layer doesn't start with N
- fail if layer has letter repetition
- fail if dimension of C =! 1 or 3

Reply:
Sorry, I see what you mean. Yeah let's already validate there!

Reply:
Added in [2fd6907](https://github.com/luxonis/luxonis-ml/pull/109/commits/2fd69074e8e55c4f160d8f28b5a9b1042a93f3a5).